### PR TITLE
`lib/posix-environ`: Environment variables

### PR DESCRIPTION
**Pull request by skuenzer**
### Base target

 - Architecture(s): [all]
 - Platform(s): [all]
 - Application(s): [all]

### Additional configuration

 - `CONFIG_LIBPOSIX_ENVIRON=y`

This PR depends on PR #867.

### Description of changes

This PR introduces the library `posix-environ` for handling environment variables. An initial set of variables can be set vva the kernel command line (`lib/uklibparam`) and/or can be compiled in.

### CODE REVIEW

**Comment by razvand** [comment]
@skuenzer, please fix the checkpatch issues.

**Comment by skuenzer** [comment]
> @skuenzer, please fix the checkpatch issues.

👍🏻  Yes, I also saw that the PR produces a warning which I'll fix with the update. In the import commit, I will switch on checkpatch ignores. I am copying unchanged code from musl and it is better to keep it in original state for tracking of changes.

**Review by jobpaardekooper — COMMENTED** [review]
I tested it locally and its working. Looked through the code and it seems good from what I can tell.

The Codacy Static Code Analysis is reporting a few new issues. Please take a look at them. It seems like at least the "The scope of the variable ... can be reduced." issues can be fixed as they are not in the code taken from musl.

File: lib/ukboot/boot.c [hunk 1]
```diff
@@ -251,12 +257,30 @@ void ukplat_entry(int argc, char *argv[])
...
-	if (unlikely(rc < 0))
-		uk_pr_crit("Failed to parse the kernel argument\n");
-	else {
-		kern_args = rc;
-		uk_pr_info("Found %d library args\n", kern_args);
+	/*
+	 * First, we scan if we can find the stop sequence in the kernel
+	 * cmdline. If not, we assume that there are no uklibparam arguments in
+	 * the command line.
+	 * NOTE: argv[0] contians the kernel/program name that we need to
```
Comment (Starnox) [inline lib/ukboot/boot.c]: Saw that the linter didn't catch this one:
s/contians/contains

**Review by Starnox — COMMENTED** [review]
It works as expected on my side! The code looks fine to me as well, just that small typo. Waiting for the checkpatch fix.

**Comment by skuenzer** [comment]
Changes are applied. Regarding Codacity: In the end only one issue applies to this PR (the others reports apply to the dependent PR #867 ). It is about `strlen` that is used in musl's original code. The issue is stated as "minor" and not really fixable with environment variables: The assumption is that they are zero terminated and there are no assumptions nor API parameters available about maximum lengths.

File: lib/ukboot/boot.c (hunk 1 above)
Comment (skuenzer) [reply lib/ukboot/boot.c]: I'll fix it with PR #867. Thanks :+1:

**Review by Starnox — APPROVED** [review]
Looks good, thanks!

Reviewed-by: Eduard-Florin Mihailescu [mihailescu.eduard@gmail.com](mailto:mihailescu.eduard@gmail.com)

**Review by DeliaPavel — APPROVED** [review]
Looks good to me!

Reviewed-by: Delia Pavel <delia_maria.pavel@stud.acs.upb.ro>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request closed without merging**